ledfx/effects/game_of_life.py

```diff
@@ -0,0 +1,319 @@
+            if self.health_check_options[HealthOptions.OSCILLATING.value]:
+                self.game.check_board_oscillating()
+
+            self.last_health_check = current_time
+
+    def step_board_if_time_elapsed(self, current_time):
+        """
+        Steps the game board if the specified time has elapsed since the last step.
+
+        Args:
+            current_time (float): The current time in seconds.
+
+        Returns:
+            None
+        """
+        if current_time - self.last_game_step >= (1 / self.base_game_speed):
+            self.game.step_board()
+            self.last_game_step = current_time
+
+    def update_image_with_board(self, image):
+        """
+        Updates the image with the current game board.
+
+        Args:
+            image (PIL.Image.Image): The original image object.
+
+        """
+        np_image = np.array(image)
+        mask = np.uint8(self.game.board)
+        np_image[mask == 1] = [255, 255, 255]
+        updated_image = Image.fromarray(np_image)
+        image.paste(updated_image, (0, 0))
+
+
+class GameOfLife:
+    """
+    Represents the Game of Life simulation.
+
+    Attributes:
+        board_history (list): A list to store the history of the game boards.
+        board_size (array): the board height and width
+        board (np array): the matrix that stores cell values
+        last_history_clear (time): the time of the last history clear
+
+    Methods:
+        initialize_board(): Initializes the game board with random cell states.
+        step_board(board): Performs one step of the game simulation.
+        board_is_dead(): Checks if the game board has reached a dead state.
+        board_is_oscillating(lookback_generations): Checks if the game board is oscillating.
+        update_image_with_board(image, board): Updates the image with the current game board.
+        cleanup_board_history(): Clears the history of the game boards.
+        add_glider(board): Generates a glider somewhere on the board.
+        add_blinker(board): Generates a blinker somewhere on the board.
+        add_toad(board): Generates a toad somewhere on the board.
+        add_beacon(board): Generates a beacon somewhere on the board.
+        add_random_entity(board): Generates a random entity on the board.
+    """
+
+    def __init__(self, height, width):
+        self.board_history = []
+        self.board_size = [height, width]
+        self.board = self.random_board()
+        _LOGGER.info("Universe invented 🌌")
+
+    def random_board(self):
+        """
+        Return a board with random cell states.
+
+        Returns:
+            numpy.ndarray: The random game board.
+        """
+        _LOGGER.info("Evolving life")
+        return np.random.choice([True, False], self.board_size)
+
+    def step_board(self):
+        """
+        Performs one step of the game simulation.
+        """
+        # _LOGGER.debug("Stepping board")
+        self.board_history.append(self.board)
```

Comment:
Just put a list length limit and remove the oldest each time once hit, no need for a depth of more that anything we leverage. I think this grows unbounded right now?

The are two ways to address history. 

Comparing histories, or the way I have done it before, each cell rather than boolean is an age value.
arbitrary values below
0 = nothing present
10 to 1, how long its been dead, starts at 10 and decrements towards 0
20 to 11, how long its been alive, starts at 20, and decrements towards 11

You can use those values to decide colors then...

Reply:
if len(self.board_history) > self.board_history_keep:
    self.board_history.pop(0)